src/packet_sources/macos.rs

```diff
@@ -0,0 +1,223 @@
+use crate::messages::{IpPacket, NetworkCommand, NetworkEvent, TunnelInfo};
+use crate::network::MAX_PACKET_SIZE;
+use crate::packet_sources::ipc::from_redirector::Message::Packet;
+use crate::packet_sources::ipc::{FromRedirector, PacketWithMeta};
+use crate::packet_sources::{ipc, PacketSourceConf, PacketSourceTask};
+use anyhow::{anyhow, Result};
+use async_trait::async_trait;
+use home::home_dir;
+use nix::{sys::stat::Mode, unistd::mkfifo};
+use prost::Message;
+use std::io::Cursor;
+use std::path::{Path, PathBuf};
+use std::process::Command;
+use tokio::net::unix::pipe;
+use tokio::sync::broadcast;
+use tokio::sync::mpsc::Sender;
+use tokio::sync::mpsc::{unbounded_channel, Receiver, UnboundedReceiver, UnboundedSender};
+
+pub const IPC_BUF_SIZE: usize = MAX_PACKET_SIZE + 4;
+
+#[allow(dead_code)]
+pub struct PipeServer {
+    ip_rx: pipe::Receiver,
+    ip_tx: pipe::Sender,
+    net_rx: pipe::Receiver,
+    net_tx: pipe::Sender,
+    filter_rx: pipe::Receiver,
+    filter_tx: pipe::Sender,
+    ip_path: PathBuf,
+    net_path: PathBuf,
+    filter_path: PathBuf,
+}
+
+impl PipeServer {
+    pub async fn new(ip_pipe: &str, net_pipe: &str, filter_pipe: &str) -> Result<Self> {
+        let home_dir = match home_dir() {
+            Some(ip_path) => ip_path,
+            None => Err(anyhow!("Failed to get home directory"))?,
+        };
+
+        let ip_path = Path::new(&home_dir).join(format!("Downloads/{}.pipe", &ip_pipe));
+        let net_path = Path::new(&home_dir).join(format!("Downloads/{}.pipe", &net_pipe));
+        let filter_path = Path::new(&home_dir).join(format!("Downloads/{}.pipe", &filter_pipe));
+
+        let (ip_rx, ip_tx) = Self::create_pipe(&ip_path)?;
+        let (net_rx, net_tx) = Self::create_pipe(&net_path)?;
+        let (filter_rx, filter_tx) = Self::create_pipe(&filter_path)?;
+
+        Ok(PipeServer {
+            ip_rx,
+            ip_tx,
+            net_rx,
+            net_tx,
+            filter_rx,
+            filter_tx,
+            ip_path,
+            net_path,
+            filter_path,
+        })
+    }
+
+    fn create_pipe(path: &PathBuf) -> Result<(pipe::Receiver, pipe::Sender)> {
+        if !path.exists() {
+            mkfifo(path, Mode::S_IRWXU)?;
+        }
+        let rx = match pipe::OpenOptions::new().unchecked(true).open_receiver(path) {
+            Ok(rx) => rx,
+            Err(e) => Err(anyhow!("Failed to open fifo receiver: {:?}", e))?,
+        };
+        let tx = match pipe::OpenOptions::new().unchecked(true).open_sender(path) {
+            Ok(tx) => tx,
+            Err(e) => Err(anyhow!("Failed to open fifo transmitter: {:?}", e))?,
+        };
```

Comment:
yeah, that drove me crazy, too. It seems in swift these pipes are handled as normal input and output files, however, if they are not already open (and saved, otherwise rust destroys the variables) then swift returns an error

Reply:
I've tried to replicate these issues at https://github.com/mhils/rust-swift-fifo, things seem to work as expected here. 
The one tricky part is waiting on the Rust side until the pipe becomes readable, see my example. Can we check this again please? :)